pkg/virtual/framework/rootapiserver/root_apiserver.go

```diff
@@ -0,0 +1,224 @@
+type InformerStarts []InformerStart
+
+type RootAPIExtraConfig struct {
+	// we phrase it like this so we can build the post-start-hook, but no one can take more indirect dependencies on informers
+	informerStart func(stopCh <-chan struct{})
+
+	VirtualWorkspaces []framework.VirtualWorkspace
+}
+
+// Validate helps ensure that we build this config correctly, because there are lots of bits to remember for now
+func (c *RootAPIExtraConfig) Validate() error {
+	ret := []error{}
+
+	return utilerrors.NewAggregate(ret)
+}
+
+type RootAPIConfig struct {
+	GenericConfig *genericapiserver.RecommendedConfig
+	ExtraConfig   RootAPIExtraConfig
+}
+
+// RootAPIServer is only responsible for serving the APIs for the virtual workspace
+// at a given root path or root path family
+// It does NOT expose oauth, related oauth endpoints, or any kube APIs.
+type RootAPIServer struct {
+	GenericAPIServer *genericapiserver.GenericAPIServer
+}
+
+type completedConfig struct {
+	GenericConfig genericapiserver.CompletedConfig
+	ExtraConfig   *RootAPIExtraConfig
+}
+
+type CompletedConfig struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*completedConfig
+}
+
+// Complete fills in any fields not set that are required to have valid data. It's mutating the receiver.
+func (c *RootAPIConfig) Complete() completedConfig {
+	cfg := completedConfig{
+		c.GenericConfig.Complete(),
+		&c.ExtraConfig,
+	}
+
+	return cfg
+}
+
+func (c *completedConfig) WithOpenAPIAggregationController(delegatedAPIServer *genericapiserver.GenericAPIServer) error {
+	return nil
+}
+
+func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget) (*RootAPIServer, error) {
+	delegateAPIServer := delegationTarget
+
+	var readys []framework.ReadyFunc
+	vwNames := sets.NewString()
+	for _, virtualWorkspace := range c.ExtraConfig.VirtualWorkspaces {
+		name := virtualWorkspace.GetName()
+		if vwNames.Has(name) {
+			return nil, errors.New("Several virtual workspaces with the same name: " + name)
+		}
+		vwNames.Insert(name)
+
+		var err error
+		delegateAPIServer, err = virtualWorkspace.Register(c.GenericConfig, delegateAPIServer)
+		if err != nil {
+			return nil, err
+		}
+		readys = append(readys, virtualWorkspace.IsReady)
+	}
+
+	c.GenericConfig.BuildHandlerChainFunc = c.getRootHandlerChain(delegateAPIServer)
+	c.GenericConfig.RequestInfoResolver = c
+	c.GenericConfig.AddHealthChecks(Readys(readys))
+
+	genericServer, err := c.GenericConfig.New("virtual-workspaces-root-apiserver", delegateAPIServer)
+	if err != nil {
+		return nil, err
+	}
+
+	s := &RootAPIServer{
+		GenericAPIServer: genericServer,
+	}
+
+	// register our poststarthooks
+	s.GenericAPIServer.AddPostStartHookOrDie("virtual-workspace-startinformers", func(context genericapiserver.PostStartHookContext) error {
+		c.ExtraConfig.informerStart(context.StopCh)
+		return nil
+	})
+
+	return s, nil
+}
+
+type Readys []framework.ReadyFunc
```

Comment:
is this used elsewhere? Does it have to be public?

Reply:
Fixed in rebased commit 9ffa9c007187b14085df141899d5d555ab725b46